src/wp-includes/fonts/class-wp-font-face.php

```diff
@@ -0,0 +1,400 @@
+		) {
+			$this->style_tag_attrs = array( 'type' => 'text/css' );
+		}
+	}
+
+	/**
+	 * Generates and prints the `@font-face` styles for the given fonts.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array[][] $fonts Optional. The font-families and their font variations.
+	 *                         See {@see wp_print_font_faces()} for the supported fields.
+	 *                         Default empty array.
+	 * }
+	 */
+	public function generate_and_print( array $fonts ) {
+		$fonts = $this->validate_fonts( $fonts );
+
+		// Bail out if there are no fonts are given to process.
+		if ( empty( $fonts ) ) {
+			return;
+		}
+
+		printf(
+			$this->get_style_element(),
+			$this->get_css( $fonts )
+		);
+	}
+
+	/**
+	 * Validates each of the font-face properties.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $fonts The fonts to valid.
+	 * @return array Prepared font-faces organized by provider and font-family.
+	 */
+	private function validate_fonts( array $fonts ) {
+		$validated_fonts = array();
+
+		foreach ( $fonts as $font_faces ) {
+			foreach ( $font_faces as $font_face ) {
+				$font_face = $this->validate_font_face_properties( $font_face );
+				// Skip if failed validation.
+				if ( false === $font_face ) {
+					continue;
+				}
+
+				$validated_fonts[] = $font_face;
+			}
+		}
+
+		return $validated_fonts;
+	}
+
+	/**
+	 * Validates each font-face property.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $font_face Font face properties to validate.
+	 * @return false|array Validated font-face on success. Else, false.
+	 */
+	private function validate_font_face_properties( array $font_face ) {
+		$font_face = wp_parse_args( $font_face, $this->font_face_property_defaults );
+
+		// Check the font-family.
+		if ( empty( $font_face['font-family'] ) || ! is_string( $font_face['font-family'] ) ) {
+			trigger_error( 'Font font-family must be a non-empty string.' );
+			return false;
+		}
+
+		// Make sure that local fonts have 'src' defined.
+		if ( empty( $font_face['src'] ) || ( ! is_string( $font_face['src'] ) && ! is_array( $font_face['src'] ) ) ) {
+			trigger_error( 'Font src must be a non-empty string or an array of strings.' );
+			return false;
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font_face['src'] ) ) {
+			foreach ( (array) $font_face['src'] as $src ) {
+				if ( empty( $src ) || ! is_string( $src ) ) {
+					trigger_error( 'Each font src must be a non-empty string.' );
+					return false;
+				}
+			}
+		}
+
+		// Check the font-weight.
+		if ( ! is_string( $font_face['font-weight'] ) && ! is_int( $font_face['font-weight'] ) ) {
+			trigger_error( 'Font font-weight must be a properly formatted string or integer.' );
+			return false;
+		}
+
+		// Check the font-display.
+		if ( ! in_array( $font_face['font-display'], $this->valid_font_display, true ) ) {
+			$font_face['font-display'] = $this->font_face_property_defaults['font-display'];
```

Comment:
Hmm, it's a doing it wrong, as an invalid value will use the default instead and the processing continues (unlike the other validation above where validation fails and it bails out).

Looking at Core, other loosely similar checks do not inform the developer either. That makes me wonder .. Should Core alert that it is using a default value if the given value fails its validation? Hmm, interesting 🤔. It might be helpful to developers to help them debug.

If this were to throw a notice, likely should wrap that notice in `empty( trim () )` to avoid the notice on an empty string value.

As this isn't a `_doing_it_wrong()` and `wp_trigger_error()` does not yet exist in Core, this helper message could be added later once `wp_trigger_error()` is introduced. It's definitely worth a good think and discussion.